trinity/protocol/eth/requests.py

```diff
@@ -40,3 +48,42 @@ def validate_response(self, response: Any) -> None:
             raise ValidationError("Response must be a tuple of `BlockHeader` objects")
 
         return self.validate_headers(cast(Tuple[BlockHeader, ...], response))
+
+
+class NodeDataRequest(BaseRequest):
+    @property
+    def max_size(self) -> int:
+        return constants.MAX_STATE_FETCH
+
+    def __init__(self, node_hashes: Tuple[Hash32, ...]) -> None:
+        self.node_hashes = node_hashes
+
+    def validate_response(self, response: Any) -> None:
+        """
+        Core `Request` API used for validation.
+        """
+        if not isinstance(response, tuple):
+            raise ValidationError("Response to `HeaderRequest` must be a tuple")
+        elif not all(isinstance(item, bytes) for item in response):
+            raise ValidationError("Response must be a tuple of `BlockHeader` objects")
+        return self.validate_nodes(cast(Tuple[bytes, ...], response))
+
+    def validate_nodes(self, nodes: Tuple[bytes, ...]) -> None:
+        if not nodes:
+            # an empty response is always valid
+            return
+
+        # TODO: move to executor since keccak is expensive
```

Comment:
I think this is very likely to cause the main thread to block for too long, thus preventing the event loop from consuming received messages fast enough which in turn will lead to us dropping peers, as we've seen recently. We should either address this as part of this PR or immediately afterwards

Reply:
Also, does this mean we're now hashing nodes twice? Once here and once in `_request_and_process_nodes()`? That'd be worrisome as this is indeed an expensive operation and we don't want to do it twice for every NodeData we receive

Reply:
> We should either address this as part of this PR or immediately afterwards

:+1: I'm going to prioritize https://github.com/ethereum/py-evm/issues/1076 today and try to get it in before merging this.

> does this mean we're now hashing nodes twice?

Yes, I'm still puzzling over what to do about this.  It's even worse in the `GetReceipts/Receipts` case because we have to generate the receipt tries to validate the response against the requested `header.receipt_root` properties.  Still not sure what I want to do about this but I know it isn't tenable to incur that overhead twice.